src/Mvc/Mvc.Core/test/Formatters/TranscodingReadStreamTest.cs

```diff
@@ -0,0 +1,113 @@
+            GetUnicodeText(TranscodingReadStream.MaxCharBufferSize);
+
+        internal static TheoryData GetLatinTextInput(int maxCharBufferSize, int maxByteBufferSize)
+        {
+            return new TheoryData<string>
+            {
+                "Hello world",
+                string.Join(string.Empty, Enumerable.Repeat("AB", 9000)),
+                new string('A', count: maxByteBufferSize),
+                new string('A', count: maxCharBufferSize),
+                new string('A', count: maxByteBufferSize + 1),
+                new string('A', count: maxCharBufferSize + 1),
+            };
+        }
+
+        internal static TheoryData GetUnicodeText(int maxCharBufferSize)
+        {
+            return new TheoryData<string>
+            {
+                new string('Æ', count: 7),
+                new string('A', count: maxCharBufferSize - 1) + 'Æ',
+                "AbĀāĂăĄąĆŊŋŌōŎŏŐőŒœŔŕŖŗŘřŚşŠšŢţŤťŦŧŨũŪūŬŭŮůŰűŲųŴŵŶŷŸŹźŻżŽžſAbc",
+               "Abcஐஒஓஔகஙசஜஞடணதநனபமயரறலளழவஷஸஹ",
+               "☀☁☂☃☄★☆☇☈☉☊☋☌☍☎☏☐☑☒☓☚☛☜☝☞☟☠☡☢☣☤☥☦☧☨☩☪☫☬☭☮☯☰☱☲☳☴☵☶☷☸",
+               new string('Æ', count: 64 * 1024),
+                new string('Æ', count: 64 * 1024 + 1),
+               "pingüino",
+                new string('ऄ', count: maxCharBufferSize + 1), // This uses 3 bytes to represent in UTF8
+            };
+        }
+
+        [Theory]
+        [MemberData(nameof(ReadAsyncInputLatin))]
+        [MemberData(nameof(ReadAsyncInputUnicode))]
+        public Task ReadAsync_Works_WhenInputIs_UTF32(string message)
+        {
+            var sourceEncoding = Encoding.UTF32;
+            return ReadAsyncTest(sourceEncoding, message);
+        }
+
+        [Theory]
+        [MemberData(nameof(ReadAsyncInputLatin))]
+        [MemberData(nameof(ReadAsyncInputUnicode))]
+        public Task ReadAsync_Works_WhenInputIs_Unicode(string message)
+        {
+            var sourceEncoding = Encoding.Unicode;
+            return ReadAsyncTest(sourceEncoding, message);
+        }
+
+        [Theory]
+        [MemberData(nameof(ReadAsyncInputLatin))]
+        [MemberData(nameof(ReadAsyncInputUnicode))]
+        public Task ReadAsync_Works_WhenInputIs_UTF7(string message)
+        {
+            var sourceEncoding = Encoding.UTF7;
+            return ReadAsyncTest(sourceEncoding, message);
+        }
+
+        [Theory]
+        [MemberData(nameof(ReadAsyncInputLatin))]
+        public Task ReadAsync_Works_WhenInputIs_WesternEuropeanEncoding(string message)
+        {
+            // Arrange
+            var sourceEncoding = Encoding.GetEncoding(28591);
+            return ReadAsyncTest(sourceEncoding, message);
+        }
+
+        [Theory]
+        [MemberData(nameof(ReadAsyncInputLatin))]
+        public Task ReadAsync_Works_WhenInputIs_ASCII(string message)
+        {
+            // Arrange
+            var sourceEncoding = Encoding.ASCII;
+            return ReadAsyncTest(sourceEncoding, message);
+        }
+
```

Comment:
I think we should definitely have a case where we use the read-stream directly. Possibly reading it character by character. I want to make sure we're exercising the special case taht requires overflow buffer.